src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs

```diff
@@ -474,58 +482,86 @@ private static bool IsValidIPv6Host(string hostText)
+                        ValidateHostPort(hostText, i + 1);
+                        return;
                     }
-                    return IsValidHostPort(hostText, i + 1);
                 }
 
                 if (!IsHex(ch) && ch != ':' && ch != '.')
                 {
-                    return false;
+                    BadHttpRequestException.Throw(RequestRejectionReason.InvalidHostHeader, hostText);
                 }
             }
 
             // Must contain a ']'
-            return false;
+            BadHttpRequestException.Throw(RequestRejectionReason.InvalidHostHeader, hostText);
         }
 
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static bool IsValidHostPort(string hostText, int offset)
+        private static void ValidateHostPort(string hostText, int offset)
         {
-            if (offset == hostText.Length)
+            // Skip bounds check for accessing the [offset] element
+            if ((uint)offset >= (uint)hostText.Length)
             {
-                return true;
+                return;
             }
 
-            if (hostText[offset] != ':' || hostText.Length == offset + 1)
+            var firstChar = hostText[offset];
+            offset++;
+            if (firstChar != ':' || (uint)offset >= (uint)hostText.Length)
             {
                 // Must have at least one number after the colon if present.
-                return false;
+                BadHttpRequestException.Throw(RequestRejectionReason.InvalidHostHeader, hostText);
             }
 
-            for (var i = offset + 1; i < hostText.Length; i++)
+            // This do+if check rather than for loop is to elimitate the bounds check, since
+            // the Jit doesn't currently pick up on it when starting at a variable offset
+            do
             {
-                if (!IsNumeric(hostText[i]))
+                // Elminate bounds check for array access
+                if ((uint)offset >= (uint)hostText.Length)
                 {
-                    return false;
+                    // Length reached, end of loop
+                    break;
                 }
-            }
 
-            return true;
+                var ch = hostText[offset];
+                offset++;
+                if (!IsNumeric(ch))
+                {
+                    BadHttpRequestException.Throw(RequestRejectionReason.InvalidHostHeader, hostText);
+                }
+            } while (true);
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         private static bool IsNumeric(char ch)
         {
-            return '0' <= ch && ch <= '9';
+            // '0' <= ch && ch <= '9'
+            // (uint)(ch - '0') <= (uint)('9' - '0')
+
+            // Subtract start of range '0'
+            // Cast to uint to change negative numbers to large numbers
+            // Check if less than 10 representing chars '0' - '9'
+            return (uint)(ch - '0') < 10u;
```

Comment:
Do we know that a subtraction then a comparison is faster than two comparisons? I ask because the earlier version certainly does read better which I guess is why you left it there as a comment. I have more or less the same question about the IsHex change.

I know you've posted some nice results from the ResponseHeadersWritingBenchmark and the Http1ConnectionBenchmark, but I'm more interested in the performance of these methods in isolation.

Reply:
Its an extra branch prediction/mis-prediction per character; rather than the extra comparison that's the issue; cpu can only run with so many branches in parallel before you hit a pipeline bubble: [Control_hazards_(branch_hazards)](https://en.wikipedia.org/wiki/Hazard_(computer_architecture)#Control_hazards_(branch_hazards))

Reply:
How do you know changing `'0' <= ch && ch <= '9'` to `(uint)(ch - '0') < 10u` improves the branch prediction?

Reply:
No branch vs 1 branch?

Reply:
Makes sense... I'd feel much more warm and fuzzy about it if there were some BenchmarkDotNet results to back up this theory.

Reply:
Have returned to my command center of my big screen monitors, so should be able to rummage something up.